### PR TITLE
Add initial ruby 3.1 support

### LINKED ISSUE #16080: Usr/lib/ruby/gems/3.1.0/gems/zeitwerk-2.5.3/lib/zeitwerk/kernel.rb:35:in `require': cannot load such file -- net/smtp (LoadError)
How to solve this ? sr/lib/ruby/gems/3.1.0/gems/zeitwerk-2.5.3/lib/zeitwerk/kernel.rb:35:in `require': cannot load such file -- net/smtp (LoadError)

### PR DESCRIPTION
**Pull request by adfoster-r7**
Adds initial Ruby 3.1 support to msfconsole

closes https://github.com/rapid7/metasploit-framework/issues/16080

### Ignored issues

1. On startup, there are `DidYouMean` warnings. These are fixed by updating to a newer version of bundler
> Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.

2. When running `exit` there are warnings from the `concurrent-ruby` gem, which is transitive dependency pinned to v1.0.5 via our active record version
```
msf6 auxiliary(scanner/http/title) > exit
./msfconsole: warning: Exception in finalizer #<Proc:0x00007f0a64b2b7e8 /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/concurrent-ruby-1.0.5/lib/concurrent/atomic/ruby_thread_local_var.rb:85>
/home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471:in `new': can't alloc thread (ThreadError)
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471:in `create_with_logging_context'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:436:in `new'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/concurrent-ruby-1.0.5/lib/concurrent/atomic/ruby_thread_local_var.rb:86:in `block in threadlocal_finalizer'
./msfconsole: warning: Exception in finalizer #<Proc:0x00007f0a64b2b950 /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/concurrent-ruby-1.0.5/lib/concurrent/atomic/ruby_thread_local_var.rb:85>
/home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471:in `new': can't alloc thread (ThreadError)
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:471:in `create_with_logging_context'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/logging-2.3.0/lib/logging/diagnostic_context.rb:436:in `new'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/concurrent-ruby-1.0.5/lib/concurrent/atomic/ruby_thread_local_var.rb:86:in `block in threadlocal_finalizer'
```

3. ~~Database support doesn't work yet, investigating if we're blocked on a Rails upgrade or if we can fix that ourselves.
Relevant links:~~ Fixed
- ~~[Stack overflow post highlighting backwards breaking changes in Psych](https://stackoverflow.com/questions/71191685/visit-psych-nodes-alias-unknown-alias-default-psychbadalias)~~
- ~~[Rails 7.x fix](https://github.com/rails/rails/pull/42249)~~
```
/opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:430:in `visit_Psych_Nodes_Alias': Unknown alias: pgsql (Psych::BadAlias)
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/visitor.rb:30:in `visit'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/visitor.rb:6:in `accept'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:35:in `accept'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:345:in `block in revive_hash'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:343:in `each'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:343:in `each_slice'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:343:in `revive_hash'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:167:in `visit_Psych_Nodes_Mapping'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/visitor.rb:30:in `visit'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/visitor.rb:6:in `accept'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:35:in `accept'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:345:in `block in revive_hash'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:343:in `each'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:343:in `each_slice'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:343:in `revive_hash'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:167:in `visit_Psych_Nodes_Mapping'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/visitor.rb:30:in `visit'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/visitor.rb:6:in `accept'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:35:in `accept'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:3[18](https://github.com/rapid7/metasploit-framework/runs/5660769299?check_suite_focus=true#step:7:18):in `visit_Psych_Nodes_Document'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/visitor.rb:30:in `visit'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/visitor.rb:6:in `accept'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:35:in `accept'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych.rb:335:in `safe_load'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych.rb:370:in `load'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych.rb:671:in `block in load_file'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych.rb:670:in `open'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/psych.rb:670:in `load_file'
	from /home/runner/work/metasploit-framework/metasploit-framework/lib/msf/core/db_manager.rb:[19](https://github.com/rapid7/metasploit-framework/runs/5660769299?check_suite_focus=true#step:7:19)4:in `init_db'
	from /home/runner/work/metasploit-framework/metasploit-framework/lib/msf/core/web_services/db_manager_proxy.rb:16:in `initialize'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/singleton.rb:127:in `new'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/singleton.rb:127:in `block in instance'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/singleton.rb:125:in `synchronize'
	from /opt/hostedtoolcache/Ruby/3.1.1/x64/lib/ruby/3.1.0/singleton.rb:125:in `instance'
	from /home/runner/work/metasploit-framework/metasploit-framework/lib/msf/core/web_services/http_db_manager_service.rb:42:in `init_db'
	from /home/runner/work/metasploit-framework/metasploit-framework/lib/msf/core/web_services/http_db_manager_service.rb:[20](https://github.com/rapid7/metasploit-framework/runs/5660769299?check_suite_focus=true#step:7:20):in `start'
	from /home/runner/work/metasploit-framework/metasploit-framework/tools/dev/msfdb_ws:80:in `<main>'
```

### Verification

- Ensure msfconsole works with Ruby 3.1
- Ensure CI passes

### CODE REVIEW

File: lib/msf/core/exploit/remote/vim_soap.rb [hunk 1]
```diff
@@ -165,7 +165,7 @@ def vim_send_soap_request(soap_data)
     if res.body.include? "NotAuthenticatedFault"
       return :expired
     elsif res.body.include? "<faultstring>"
-      @vim_soap_error = res.body.match(/<faultstring>([^\c ]+?)<\/faultstring>/)[1]
+      @vim_soap_error = res.body.match(/<faultstring>(.+?)<\/faultstring>/m)[1]
```
Comment (adfoster-r7) [inline lib/msf/core/exploit/remote/vim_soap.rb:168]: Fixing the warning:

```
user@ubuntu:/mnt/hgfs/metasploit-framework$ bundle exec ruby ./msfconsole -q
/mnt/hgfs/metasploit-framework/lib/msf/core/exploit/remote/vim_soap.rb:168: warning: invalid character syntax; use \C-\s
[*] Using configured payload windows/meterpreter/reverse_tcp
msf6 exploit(windows/smb/psexec) > 
```

File: metasploit-framework.gemspec [hunk 2]
```diff
@@ -142,6 +142,7 @@ Gem::Specification.new do |spec|
   spec.add_runtime_dependency 'bcrypt_pbkdf'
   spec.add_runtime_dependency 'ruby_smb', '~> 3.0'
   spec.add_runtime_dependency 'net-ldap'
+  spec.add_runtime_dependency 'net-smtp'
```
Comment (adfoster-r7) [inline metasploit-framework.gemspec:145]: Fixing the startup crash caused by net-smtp no longer being part of Ruby by default

> > The following default gems are now bundled gems. You need to add the following libraries to Gemfile under the > bundler environment.
> >   - net-ftp 0.1.3
> >  - net-imap 0.2.2
> >  - net-pop 0.1.1
> >  - net-smtp 0.3.1
> 
> https://www.ruby-lang.org/en/news/2021/12/25/ruby-3-1-0-released/

```
/home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/kernel.rb:35:in `require': cannot load such file -- net/smtp (LoadError)
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/kernel.rb:35:in `require'
	from /mnt/hgfs/metasploit-framework/lib/rex/proto/sms/model.rb:29:in `<top (required)>'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/kernel.rb:27:in `require'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/kernel.rb:27:in `require'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:95:in `const_get'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:95:in `cget'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:237:in `block (2 levels) in eager_load'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:26:in `block in ls'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:18:in `each_child'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader/helpers.rb:18:in `ls'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:232:in `block in eager_load'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:217:in `synchronize'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:217:in `eager_load'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:317:in `each'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/zeitwerk-2.5.4/lib/zeitwerk/loader.rb:317:in `eager_load_all'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/railties-6.1.5/lib/rails/application/finisher.rb:133:in `block in <module:Finisher>'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/railties-6.1.5/lib/rails/initializable.rb:32:in `instance_exec'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/railties-6.1.5/lib/rails/initializable.rb:32:in `run'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/railties-6.1.5/lib/rails/initializable.rb:61:in `block in run_initializers'
	from /home/user/.rvm/rubies/ruby-3.1.1/lib/ruby/3.1.0/tsort.rb:228:in `block in tsort_each'
	from /home/user/.rvm/rubies/ruby-3.1.1/lib/ruby/3.1.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
	from /home/user/.rvm/rubies/ruby-3.1.1/lib/ruby/3.1.0/tsort.rb:431:in `each_strongly_connected_component_from'
	from /home/user/.rvm/rubies/ruby-3.1.1/lib/ruby/3.1.0/tsort.rb:349:in `block in each_strongly_connected_component'
	from /home/user/.rvm/rubies/ruby-3.1.1/lib/ruby/3.1.0/tsort.rb:347:in `each'
	from /home/user/.rvm/rubies/ruby-3.1.1/lib/ruby/3.1.0/tsort.rb:347:in `call'
	from /home/user/.rvm/rubies/ruby-3.1.1/lib/ruby/3.1.0/tsort.rb:347:in `each_strongly_connected_component'
	from /home/user/.rvm/rubies/ruby-3.1.1/lib/ruby/3.1.0/tsort.rb:226:in `tsort_each'
	from /home/user/.rvm/rubies/ruby-3.1.1/lib/ruby/3.1.0/tsort.rb:205:in `tsort_each'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/railties-6.1.5/lib/rails/initializable.rb:60:in `run_initializers'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/railties-6.1.5/lib/rails/application.rb:391:in `initialize!'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/railties-6.1.5/lib/rails/railtie.rb:207:in `public_send'
	from /home/user/.rvm/gems/ruby-3.1.1@metasploit-framework/gems/railties-6.1.5/lib/rails/railtie.rb:207:in `method_missing'
	from /mnt/hgfs/metasploit-framework/config/environment.rb:4:in `<top (required)>'
	from /mnt/hgfs/metasploit-framework/lib/msfenv.rb:17:in `require'
	from /mnt/hgfs/metasploit-framework/lib/msfenv.rb:17:in `<top (required)>'
	from ./msfconsole:18:in `require'
	from ./msfconsole:18:in `<main>'
```

**Comment by adfoster-r7** [comment]
Looks like a real test failure to me; I believe Ruby 3.1.1 changes error messages 🤔 

https://rubyreferences.github.io/rubychanges/3.1.html#standard-library-content-changes

Ruby 3
```
msf6 auxiliary(scanner/http/title) > check 123.123.123.123
[-] Check failed: NoMethodError This module does not support check.
```

Ruby 3.1.1
```
msf6 auxiliary(scanner/http/title) > check 123.123.123.123
[-] Check failed: NoMethodError This module does not support check.

      raise ::NoMethodError.new(Msf::Exploit::CheckCode::Unsupported.message, 'check')
      ^^^^^
```

File: lib/msfenv.rb [hunk 3]
```diff
@@ -18,4 +18,15 @@
...
+# Disable the enhanced error messages introduced as part of Ruby 3.1, as some error messages are directly shown to users,
+# and the default ErrorHighlight formatter displays unneeded Ruby code to the user
+# https://github.com/ruby/error_highlight/tree/f3626b9032bd1024d058984329accb757687cee4#custom-formatter
+if defined?(ErrorHighlight)
+  formatter = Object.new
+  def formatter.message_for(_spot)
+    ""
+  end
+  ErrorHighlight.formatter = formatter
+end
```
Comment (adfoster-r7) [inline lib/msfenv.rb:30]: Error changes:
https://rubyreferences.github.io/rubychanges/3.1.html#standard-library-content-changes

Ruby 3
```
msf6 auxiliary(scanner/http/title) > check 123.123.123.123
[-] Check failed: NoMethodError This module does not support check.
```

Ruby 3.1.1
```
msf6 auxiliary(scanner/http/title) > check 123.123.123.123
[-] Check failed: NoMethodError This module does not support check.

      raise ::NoMethodError.new(Msf::Exploit::CheckCode::Unsupported.message, 'check')
      ^^^^^
```

File: lib/msf/core/exploit/remote/kerberos/client/tgs_request.rb [hunk 4]
```diff
@@ -225,9 +225,9 @@ def build_subkey(opts={})
             # @see Rex::Proto::Kerberos::Model::KdcRequestBody
             def build_tgs_request_body(opts = {})
               options = opts[:options] || 0x50800000 # Forwardable, Proxiable, Renewable
-              from = opts[:from] || Time.utc('1970-01-01-01 00:00:00')
-              till = opts[:till] || Time.utc('1970-01-01-01 00:00:00')
-              rtime = opts[:rtime] || Time.utc('1970-01-01-01 00:00:00')
+              from = opts[:from] || Time.at(0)
+              till = opts[:till] || Time.at(0)
+              rtime = opts[:rtime] || Time.at(0)
```
Comment (adfoster-r7) [inline lib/msf/core/exploit/remote/kerberos/client/tgs_request.rb:230]:
```
3.1.1 :001 > Time.utc('1970-01-01-01 00:00:00')
(irb):1:in `utc': invalid value for Integer(): "1970-01-01-01 00:00:00" (ArgumentError)
	from (irb):1:in `<main>'
	from /Users/user/.rvm/rubies/ruby-3.1.1/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
	from /Users/user/.rvm/gems/ruby-3.1.1@metasploit-framework/bin/irb:25:in `load'
	from /Users/user/.rvm/gems/ruby-3.1.1@metasploit-framework/bin/irb:25:in `<main>'
	from /Users/user/.rvm/gems/ruby-3.1.1@metasploit-framework/bin/ruby_executable_hooks:22:in `eval'
	from /Users/user.rvm/gems/ruby-3.1.1@metasploit-framework/bin/ruby_executable_hooks:22:in `<main>
```

**Comment by dwelch-r7** [comment]
# Release Notes

Adds initial support for ruby 3.1

**Pull request merged**